### PR TITLE
Fixed bugs that caused crashes when new attributes were added

### PR DESCRIPTION
**Pull request by MarkRivers**
Fixed bugs that caused crashes when new attributes were added and fields changed.

### CODE REVIEW

File: ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp [hunk 1]
```diff
@@ -90,8 +90,11 @@ NDColorMode_t NTNDArrayConverter::getColorMode (void)
     {
         if((*it)->getSubField<PVString>("name")->get() == "ColorMode")
```
Comment (brunoseivam) [inline ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp]: Maybe this should be `getSubFieldT` as well for consistency?

Comment (MarkRivers) [reply ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp]: I've done that now.  Please take a look at the new version.

**Comment by brunoseivam** [comment]
It looks good to me, but I haven't tested it. It would be nice if we could use `getAs` in some places: for example, `ColorMode` is not _strictly required_ to be an `int32`, it could be any integer, or string representation of an integer:

``` C++
PVScalarPtr valueFld(destUnion->get<PVScalar>());
if(!valueFld)
    throw std::runtime_error("Empty or non-scalar ColorMode attribute");
int cm = valueFld->getAs<pvInt>();
```

`getAs` would do the conversion automatically, if I understand it correctly.

I do want to, at some point, revisit the whole converter code and make it nicer.

**Comment by MarkRivers** [comment]
getColorMode is only called when converting from NTNDArray to NDArray.  I have tested it by running pvaDriver and changing the color mode on the pva server.  

I also tested that it fails gracefully if ColorMode is the wrong type by temporarily changing the test from PVInt to PVLong.

**Comment by MarkRivers** [comment]
I am interested in any comments from @brunoseivam or @mdavidsaver about line 527

    std::tr1::shared_ptr<pvAttrType> valueFld(destUnion->get<pvAttrType>());

Is there a cleaner way to do this?  It is hard-coding information about the implementation of PVIntPtr for example as an std::tr1::shared_ptr.  Is there a way to avoid this?

**Comment by brunoseivam** [comment]
I am not 100% sure, but you could probably do `pvAttrType::shared_pointer` instead

File: ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp [hunk 2]
```diff
@@ -517,12 +523,13 @@ void NTNDArrayConverter::fromAttribute (PVStructurePtr dest, NDAttribute *src)
...
     src->getValue(src->getDataType(), (void*)&value);
 
-    PVUnionPtr destUnion(dest->getSubField<PVUnion>("value"));
-
-    if(!destUnion->get())
-        destUnion->set(PVDC->createPVScalar<pvAttrType>());
-
-    static_pointer_cast<pvAttrType>(destUnion->get())->put(value);
+    PVUnionPtr destUnion(dest->getSubFieldT<PVUnion>("value"));
+    std::tr1::shared_ptr<pvAttrType> valueFld(destUnion->get<pvAttrType>());
```
[the remark this answers is not in the document]
Comment (MarkRivers) [reply ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp]: I've made this change, thanks.

File: ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp [hunk 3]
```diff
@@ -88,10 +88,16 @@ NDColorMode_t NTNDArrayConverter::getColorMode (void)
...
-        if((*it)->getSubField<PVString>("name")->get() == "ColorMode")
+        PVStringPtr nameFld((*it)->getSubFieldT<PVString>("name"));
+        if(!nameFld)
+                throw std::runtime_error("name attribute is not of type PVString");
+        if(nameFld->get() == "ColorMode")
         {
-            PVUnionPtr field((*it)->getSubField<PVUnion>("value"));
-            int cm = static_pointer_cast<PVInt>(field->get())->get();
+            PVUnionPtr valueUnion((*it)->getSubFieldT<PVUnion>("value"));
+            PVIntPtr valueFld(valueUnion->get<PVInt>());
```
[the remark this answers is not in the document]
Comment (MarkRivers) [reply ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp]: I need some help on the syntax for this, I couldn't figure it out.

File: ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp [hunk 4]
```diff
@@ -88,10 +88,16 @@ NDColorMode_t NTNDArrayConverter::getColorMode (void)
     for(PVStructureArray::const_svector::iterator it(attrs.cbegin());
             it != attrs.cend(); ++it)
     {
-        if((*it)->getSubField<PVString>("name")->get() == "ColorMode")
+        PVStringPtr nameFld((*it)->getSubFieldT<PVString>("name"));
+        if(!nameFld)
+                throw std::runtime_error("name attribute is not of type PVString");
```
[the remark this answers is not in the document]
Comment (MarkRivers) [reply ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp]: I removed that test, and verified that getSubFieldT will itself throw an exception if the subfield has the wrong type (by temporarily changing PVString to PVInt).

File: ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdavidsaver) [reply ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp]: Something like:

```c++
int mode = 0;
PVScalar::shared_pointer valueFld(valueUnion->get<PVScalar>());
if(valueFld) {
    mode = valueFld->getAs<int32>();
}
```

[the remark this answers is not in the document]
Comment (MarkRivers) [reply ADApp/ntndArrayConverterSrc/ntndArrayConverter.cpp]: Got it.  I've made the change and tested it.

**Pull request merged**